lib/events.ex

```diff
@@ -1,3 +1,104 @@
 defmodule Events do
-  @moduledoc false
+  @moduledoc """
+  Events allows to attach functions to be called when events are emitted
+
+  Note that all subscribed functions are called in the process which emits the event.
+  """
+
+  require Logger
+
+  @callback_mod Application.fetch_env!(:events, :impl)
+
+  @type handler_id :: term()
+  @type event_name :: [atom()]
+  @type event_value :: number()
+  @type event_prefix :: [atom()]
+
+  @doc """
+  Attaches a handler to the event prefix.
+
+  `handler_id` must be unique, if another handler with the same ID already exists the
+  `:error` atom is be returned.
+
+  When events with `event_prefix` are emitted, function `function` in module `module` will be
+  called with three arguments:
+  * the exact event name (see `execute/2`)
+  * the event value (see `execute/2`)
+  * the handler configuration, `config`, or `nil` if one wasn't provided
+
+  If the function fails (raises, exits or throws) then the handler is removed.
+  """
+  @spec attach(handler_id, event_prefix, module, function :: atom) ::
+          :ok | {:error, :already_exists}
+  @spec attach(handler_id, event_prefix, module, function :: atom, config :: term) ::
+          :ok | {:error, :already_exists}
+  def attach(handler_id, event_prefix, module, function, config \\ nil) do
+    assert_event_name_or_prefix(event_prefix)
+    @callback_mod.attach(handler_id, event_prefix, module, function, config)
+  end
+
+  @doc """
+  Removes existing handler.
+
+  If the handler with given ID doesn't exist, `:error` is returned.
+  """
+  @spec detach(handler_id) :: :ok | {:error, :doesnt_exist}
```

Comment:
What about `:not_found` instead of `:doesnt_exist`. `:not_found` is more common in APIs. Is what a supervisor would return too.

Reply:
Makes sense! :) I'll fixup the commit.